File: ios/CameraViewOldManagerImpl.swift
//
//  CameraViewOldManagerImpl.swift
//  VisionCameraOld
//
//  Created by Marc Rousavy on 09.11.20.
//  Copyright © 2020 mrousavy. All rights reserved.
//

import AVFoundation
import Foundation

@objc
public class CameraViewOldManagerImpl: NSObject {
  private weak var bridge: RCTBridge?
  private var runtimeManager: FrameProcessorRuntimeManagerOld?

  @objc
  public init(bridge: RCTBridge) {
    self.bridge = bridge
    super.init()

    // Install Frame Processor bindings and setup Runtime
    if VISION_CAMERA_ENABLE_FRAME_PROCESSORS {
      CameraQueues.frameProcessorQueue.async {
        self.runtimeManager = FrameProcessorRuntimeManagerOld(bridge: bridge)
        bridge.runOnJS {
          self.runtimeManager!.installFrameProcessorBindings()
        }
      }
    }
  }

  @objc
  public func createView() -> UIView {
    return CameraViewOld()
  }

  // MARK: - Property Setters

  @objc
  public func setIsActive(_ view: CameraViewOld, value: Bool) {
    view.isActive = value
  }

  @objc
  public func setCameraId(_ view: CameraViewOld, value: String?) {
    view.cameraId = value as NSString?
  }

  @objc
  public func setEnableDepthData(_ view: CameraViewOld, value: Bool) {
    view.enableDepthData = value
  }

  @objc
  public func setEnableHighQualityPhotos(_ view: CameraViewOld, value: NSNumber?) {
    view.enableHighQualityPhotos = value
  }

  @objc
  public func setEnablePortraitEffectsMatteDelivery(_ view: CameraViewOld, value: Bool) {
    view.enablePortraitEffectsMatteDelivery = value
  }

  @objc
  public func setPhoto(_ view: CameraViewOld, value: NSNumber?) {
    view.photo = value
  }

  @objc
  public func setVideo(_ view: CameraViewOld, value: NSNumber?) {
    view.video = value
  }

  @objc
  public func setAudio(_ view: CameraViewOld, value: NSNumber?) {
    view.audio = value
  }

  @objc
  public func setEnableFrameProcessor(_ view: CameraViewOld, value: Bool) {
    view.enableFrameProcessor = value
  }

  @objc
  public func setFormat(_ view: CameraViewOld, value: NSDictionary?) {
    view.format = value
  }

  @objc
  public func setFps(_ view: CameraViewOld, value: NSNumber?) {
    view.fps = value
  }

  @objc
  public func setFrameProcessorFps(_ view: CameraViewOld, value: NSNumber) {
    view.frameProcessorFps = value
  }

  @objc
  public func setHdr(_ view: CameraViewOld, value: NSNumber?) {
    view.hdr = value
  }

  @objc
  public func setLowLightBoost(_ view: CameraViewOld, value: NSNumber?) {
    view.lowLightBoost = value
  }

  @objc
  public func setColorSpace(_ view: CameraViewOld, value: String?) {
    view.colorSpace = value as NSString?
  }

  @objc
  public func setVideoStabilizationMode(_ view: CameraViewOld, value: String?) {
    view.videoStabilizationMode = value as NSString?
  }

  @objc
  public func setPreset(_ view: CameraViewOld, value: String?) {
    view.preset = value
  }

  @objc
  public func setTorch(_ view: CameraViewOld, value: String) {
    view.torch = value
  }

  @objc
  public func setZoom(_ view: CameraViewOld, value: NSNumber) {
    view.zoom = value
  }

  @objc
  public func setEnableZoomGesture(_ view: CameraViewOld, value: Bool) {
    view.enableZoomGesture = value
  }

  @objc
  public func setOrientation(_ view: CameraViewOld, value: String?) {
    view.orientation = value as NSString?
  }

  // MARK: - Event Handlers

  @objc
  public func setOnError(_ view: CameraViewOld, value: RCTDirectEventBlock?) {
    view.onError = value
  }

  @objc
  public func setOnInitialized(_ view: CameraViewOld, value: RCTDirectEventBlock?) {
    view.onInitialized = value
  }

  @objc
  public func setOnFrameProcessorPerformanceSuggestionAvailable(_ view: CameraViewOld, value: RCTDirectEventBlock?) {
    view.onFrameProcessorPerformanceSuggestionAvailable = value
  }

  @objc
  public func setOnViewReady(_ view: CameraViewOld, value: RCTDirectEventBlock?) {
    view.onViewReady = value
  }
}

File: android/src/main/java/com/mrousavy/old/camera/CameraViewOldManagerImpl.kt
package com.mrousavy.old.camera

import android.view.View
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.annotations.ReactProp

class CameraViewOldManagerImpl(private val reactContext: ReactApplicationContext) {

  private val cameraViewTransactions = mutableMapOf<CameraViewOld, ArrayList<String>>()

  fun createViewInstance(context: ThemedReactContext): CameraViewOld {
    val cameraViewModule = context.getNativeModule(CameraViewOldModule::class.java)
      ?: context.getNativeModule(CameraViewOldTurboModule::class.java)

    return if (cameraViewModule is CameraViewOldModule) {
      CameraViewOld(context, cameraViewModule.moduleImpl.frameProcessorThread)
    } else {
      val turboModule = cameraViewModule as CameraViewOldTurboModule
      CameraViewOld(context, turboModule.moduleImpl.frameProcessorThread)
    }
  }

  fun onAfterUpdateTransaction(view: CameraViewOld) {
    val changedProps = cameraViewTransactions[view] ?: ArrayList()
    view.update(changedProps)
    cameraViewTransactions.remove(view)
  }

  fun getExportedCustomDirectEventTypeConstants(): Map<String, Any>? {
    return mapOf(
      "cameraViewReady" to mapOf("registrationName" to "onViewReady"),
      "cameraInitialized" to mapOf("registrationName" to "onInitialized"),
      "cameraError" to mapOf("registrationName" to "onError"),
      "cameraPerformanceSuggestionAvailable" to mapOf("registrationName" to "onFrameProcessorPerformanceSuggestionAvailable")
    )
  }

  private fun addChangedPropToTransaction(view: CameraViewOld, changedProp: String) {
    if (cameraViewTransactions[view] == null) {
      cameraViewTransactions[view] = ArrayList()
    }
    cameraViewTransactions[view]!!.add(changedProp)
  }

  fun setCameraId(view: CameraViewOld, cameraId: String?) {
    if (view.cameraId != cameraId) {
      addChangedPropToTransaction(view, "cameraId")
    }
    view.cameraId = cameraId
  }

  fun setIsActive(view: CameraViewOld, isActive: Boolean) {
    if (view.isActive != isActive) {
      addChangedPropToTransaction(view, "isActive")
    }
    view.isActive = isActive
  }

  fun setPhoto(view: CameraViewOld, photo: Boolean?) {
    if (view.photo != photo) {
      addChangedPropToTransaction(view, "photo")
    }
    view.photo = photo
  }

  fun setVideo(view: CameraViewOld, video: Boolean?) {
    if (view.video != video) {
      addChangedPropToTransaction(view, "video")
    }
    view.video = video
  }

  fun setAudio(view: CameraViewOld, audio: Boolean?) {
    if (view.audio != audio) {
      addChangedPropToTransaction(view, "audio")
    }
    view.audio = audio
  }

  fun setEnableFrameProcessor(view: CameraViewOld, enableFrameProcessor: Boolean) {
    if (view.enableFrameProcessor != enableFrameProcessor) {
      addChangedPropToTransaction(view, "enableFrameProcessor")
    }
    view.enableFrameProcessor = enableFrameProcessor
  }

  fun setFrameProcessorFps(view: CameraViewOld, frameProcessorFps: Double) {
    if (view.frameProcessorFps != frameProcessorFps) {
      addChangedPropToTransaction(view, "frameProcessorFps")
    }
    view.frameProcessorFps = frameProcessorFps
  }

  fun setFormat(view: CameraViewOld, format: ReadableMap?) {
    if (view.format != format) {
      addChangedPropToTransaction(view, "format")
    }
    view.format = format
  }

  fun setFps(view: CameraViewOld, fps: Int) {
    if (view.fps != fps) {
      addChangedPropToTransaction(view, "fps")
    }
    view.fps = fps
  }

  fun setHdr(view: CameraViewOld, hdr: Boolean?) {
    if (view.hdr != hdr) {
      addChangedPropToTransaction(view, "hdr")
    }
    view.hdr = hdr
  }

  fun setLowLightBoost(view: CameraViewOld, lowLightBoost: Boolean?) {
    if (view.lowLightBoost != lowLightBoost) {
      addChangedPropToTransaction(view, "lowLightBoost")
    }
    view.lowLightBoost = lowLightBoost
  }

  fun setColorSpace(view: CameraViewOld, colorSpace: String?) {
    if (view.colorSpace != colorSpace) {
      addChangedPropToTransaction(view, "colorSpace")
    }
    view.colorSpace = colorSpace
  }

  fun setVideoStabilizationMode(view: CameraViewOld, videoStabilizationMode: String?) {
    if (view.videoStabilizationMode != videoStabilizationMode) {
      addChangedPropToTransaction(view, "videoStabilizationMode")
    }
    view.videoStabilizationMode = videoStabilizationMode
  }

  fun setPreset(view: CameraViewOld, preset: String?) {
    if (view.preset != preset) {
      addChangedPropToTransaction(view, "preset")
    }
    view.preset = preset
  }

  fun setTorch(view: CameraViewOld, torch: String?) {
    if (view.torch != torch) {
      addChangedPropToTransaction(view, "torch")
    }
    view.torch = torch
  }

  fun setZoom(view: CameraViewOld, zoom: Double) {
    if (view.zoom != zoom) {
      addChangedPropToTransaction(view, "zoom")
    }
    view.zoom = zoom
  }

  fun setEnableDepthData(view: CameraViewOld, enableDepthData: Boolean) {
    if (view.enableDepthData != enableDepthData) {
      addChangedPropToTransaction(view, "enableDepthData")
    }
    view.enableDepthData = enableDepthData
  }

  fun setEnableHighQualityPhotos(view: CameraViewOld, enableHighQualityPhotos: Boolean?) {
    if (view.enableHighQualityPhotos != enableHighQualityPhotos) {
      addChangedPropToTransaction(view, "enableHighQualityPhotos")
    }
    view.enableHighQualityPhotos = enableHighQualityPhotos
  }

  fun setEnablePortraitEffectsMatteDelivery(view: CameraViewOld, enablePortraitEffectsMatteDelivery: Boolean) {
    if (view.enablePortraitEffectsMatteDelivery != enablePortraitEffectsMatteDelivery) {
      addChangedPropToTransaction(view, "enablePortraitEffectsMatteDelivery")
    }
    view.enablePortraitEffectsMatteDelivery = enablePortraitEffectsMatteDelivery
  }

  fun setEnableZoomGesture(view: CameraViewOld, enableZoomGesture: Boolean) {
    if (view.enableZoomGesture != enableZoomGesture) {
      addChangedPropToTransaction(view, "enableZoomGesture")
    }
    view.enableZoomGesture = enableZoomGesture
  }

  fun setOrientation(view: CameraViewOld, orientation: String?) {
    if (view.orientation != orientation) {
      addChangedPropToTransaction(view, "orientation")
    }
    view.orientation = orientation
  }
}
